[Discuss] Windows build failed due to uint64_t

We recently got the compilation error [C2398](https://docs.microsoft.com/en-us/cpp/error-messages/compiler-errors-1/compiler-error-c2397?view=vs-2019) (interestingly, it `is` C2398, not C2397, as listed [here](https://docs.microsoft.com/en-us/cpp/error-messages/compiler-errors-1/compiler-errors-c2300-through-c2399?view=vs-2019)) during nightly build on Windows platform.
```
\\local_config_tf\include\unsupported\Eigen\CXX11\src/Tensor/TensorMap.h(267): error C2398: Element '3': conversion from 'uint64_t' to 'ptrdiff_t' requires a narrowing conversion
tensorflow_quantum/core/ops/tfq_calculate_unitary_op.cc(106): note: see reference to function template instantiation 'std::complex<float> &Eigen::TensorMap<Eigen::Tensor<T,3,1,IndexType>,16,Eigen::MakePointer>::operator ()<uint64_t>(__int64,__int64,uint64_t)' being compiled
        with
        [
            T=std::complex<float>,
            IndexType=Eigen::DenseIndex
        ]
```

The ln 106 in tfq_calculate_unitary_op.cc is [this](https://github.com/tensorflow/quantum/blob/master/tensorflow_quantum/core/ops/tfq_calculate_unitary_op.cc#L106).
```cc
tensorflow::Tensor *output = nullptr;
...
auto output_tensor = output->tensor<std::complex<float>, 3>();
...
for (uint64_t j = 0; j < state_size; j++) {
  for (uint64_t k = 0; k < state_size; k++) {
    output_tensor(i, j, k) = state->GetEntry(j, k);
  }
}
```
The remedy in MSDN docs is specifying the type casting, e.g. `static_cast<ptrdiff_t>(j)` and  `static_cast<ptrdiff_t>(k)` in my understanding.  In addition, when I looked up the definition of tensorflow::Tensor::tensor(), I found out:
```cc
template <typename T, size_t NDIMS>
typename TTypes<T, NDIMS>::Tensor tensor();
```
I am curious about if `static_cast<size_t>()` is also possible or not. However, we really want to take 64-bit address due to the big size of tensors, but the fact that `size_t` can be 32-bit, dependent of platforms, and I am worrying if it is intended behavior or not.

I hope to hear your ideas. Thanks!

--- a/tensorflow_quantum/core/ops/tfq_calculate_unitary_op.cc
+++ b/tensorflow_quantum/core/ops/tfq_calculate_unitary_op.cc
@@ -103,21 +103,27 @@ class TfqSimulateStateOp : public tensorflow::OpKernel {
         uint64_t state_size = uint64_t(1) << state->GetNumQubits();
         for (uint64_t j = 0; j < state_size; j++) {
           for (uint64_t k = 0; k < state_size; k++) {
-            output_tensor(i, j, k) = state->GetEntry(j, k);
+            // Cast to size_t to keep windows compiler happy.
+            // We run less of a risk of overflowing size_t since
+            // this is a unitary and not a state.
+            output_tensor(static_cast<size_t>(i), static_cast<size_t>(j),
+                          static_cast<size_t>(k)) = state->GetEntry(j, k);
           }
         }
         // -2 padding for lower portion.
         for (uint64_t j = state_size; j < (uint64_t(1) << max_num_qubits);
              j++) {
           for (uint64_t k = 0; k < (uint64_t(1) << max_num_qubits); k++) {
-            output_tensor(i, j, k) = std::complex<float>(-2, 0);
+            output_tensor(static_cast<size_t>(i), static_cast<size_t>(j),
+                          static_cast<size_t>(k)) = std::complex<float>(-2, 0);
           }
         }
         // -2 padding for right portion.
         for (uint64_t j = 0; j < state_size; j++) {
           for (uint64_t k = state_size; k < (uint64_t(1) << max_num_qubits);
                k++) {
-            output_tensor(i, j, k) = std::complex<float>(-2, 0);
+            output_tensor(static_cast<size_t>(i), static_cast<size_t>(j),
+                          static_cast<size_t>(k)) = std::complex<float>(-2, 0);
           }
         }
         old_num_qubits = num;

